google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerCloudMonitoringExporter.java

```diff
@@ -0,0 +1,246 @@
+      logger.log(Level.WARNING, "Metric data has different a projectId. Skip exporting.");
+      return CompletableResultCode.ofFailure();
+    }
+
+    List<TimeSeries> spannerTimeSeries;
+    try {
+      spannerTimeSeries =
+          SpannerCloudMonitoringExporterUtils.convertToSpannerTimeSeries(spannerMetricData);
+    } catch (Throwable e) {
+      logger.log(
+          Level.WARNING,
+          "Failed to convert spanner metric data to cloud monitoring timeseries.",
+          e);
+      return CompletableResultCode.ofFailure();
+    }
+
+    ProjectName projectName = ProjectName.of(spannerProjectId);
+
+    ApiFuture<List<Empty>> futureList = exportTimeSeriesInBatch(projectName, spannerTimeSeries);
+
+    CompletableResultCode spannerExportCode = new CompletableResultCode();
+    ApiFutures.addCallback(
+        futureList,
+        new ApiFutureCallback<List<Empty>>() {
+          @Override
+          public void onFailure(Throwable throwable) {
+            if (spannerExportFailureLogged.compareAndSet(false, true)) {
+              String msg = "createServiceTimeSeries request failed for spanner metrics.";
+              if (throwable instanceof PermissionDeniedException) {
+                // TODO: Add the link of public documentation when available in the log message.
+                msg +=
+                    String.format(
+                        " Need monitoring metric writer permission on project=%s.",
+                        projectName.getProject());
+              }
+              logger.log(Level.WARNING, msg, throwable);
+            }
+            spannerExportCode.fail();
+          }
+
+          @Override
+          public void onSuccess(List<Empty> empty) {
+            // When an export succeeded reset the export failure flag to false so if there's a
+            // transient failure it'll be logged.
+            spannerExportFailureLogged.set(false);
+            spannerExportCode.succeed();
+          }
+        },
+        MoreExecutors.directExecutor());
+
+    return spannerExportCode;
+  }
+
+  private ApiFuture<List<Empty>> exportTimeSeriesInBatch(
+      ProjectName projectName, List<TimeSeries> timeSeries) {
+    List<ApiFuture<Empty>> batchResults = new ArrayList<>();
+
+    for (List<TimeSeries> batch : Iterables.partition(timeSeries, EXPORT_BATCH_SIZE_LIMIT)) {
+      CreateTimeSeriesRequest req =
+          CreateTimeSeriesRequest.newBuilder()
+              .setName(projectName.toString())
+              .addAllTimeSeries(batch)
+              .build();
+      ApiFuture<Empty> f = this.client.createServiceTimeSeriesCallable().futureCall(req);
+      batchResults.add(f);
+    }
+
+    return ApiFutures.allAsList(batchResults);
+  }
+
+  @Override
+  public CompletableResultCode flush() {
+    if (lastExportCode != null) {
+      return lastExportCode;
+    }
+    return CompletableResultCode.ofSuccess();
+  }
```

Comment:
The usual use of this would be if the exporter does additional asynchronous queueing or retries, `flush()` would drain the queue or wait for retries to finish. I imagine this could just be an empty method here, which is what we do for the user facing exporter https://github.com/GoogleCloudPlatform/opentelemetry-operations-java/blob/11c85b9e3809a60a6348982ee9ea38d28cbe49db/exporters/metrics/src/main/java/com/google/cloud/opentelemetry/metric/InternalMetricExporter.java#L261-L269

Reply:
User facing exporter also returns [CompletableResultCode.ofSuccess()](https://github.com/GoogleCloudPlatform/opentelemetry-operations-java/blob/11c85b9e3809a60a6348982ee9ea38d28cbe49db/exporters/metrics/src/main/java/com/google/cloud/opentelemetry/metric/InternalMetricExporter.java#L261-L269) . We have just added an extra check if the `lastExportCode` has any code set

Reply:
Yup. I just was concerned about semantics. If the previous export failed say 30 seconds ago, why would we indicate the force flush failed as well?

Reply:
Could you explain the purpose of that check ?

The [`flush()`](https://github.com/open-telemetry/opentelemetry-java/blob/0aacc55d1e3f5cc6dbb4f8fa26bcb657b01a7bc9/sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/export/MetricExporter.java#L64) method is an hint to the exporter to complete export for any metrics previously collected - and return the result of the flush operation.

This check would cause to violate this.

Reply:
Removed this check